### DISCUSSION TITLE
About Group: generate,update

### DISCUSSION THREAD

**Discussion by streetycat** [buckyos/CYFS#118] [Ideas]
The `Group` object is mutable; we will update it regularly. So, we must use blockchain to update it to ensure version synchronization and verifiability.

Now, there is my solution, anyone can suggest for me.

1. There are 2 role in Group:
  * Member: They are almost readonly except to update the self infomation.
  * Admin: They can update the infomations according to authorization.
 
 2. Configure permissions
 
   * Assign each information a category.
   * Define the weight threshold required for each information updates
   * Grant each administrator a voting weight for each category
   * Update the latest version when the update collect enough votes

3. Information classification

<table>
	<tr>
		<th>Category resume</th>
		<th>Category</th>
		<th>Field</th>
		<th>Notes</th>
	</tr>
    <tr>
		<td rowspan="2">Member list</td>
        <td>admin</td>
        <td>admins</td>
        <td>add/remove</td>
	</tr>
    <tr>
        <td>member</td>
        <td>members</td>
        <td>add/remove</td>
	</tr>
    <tr>
		<td rowspan="4">Display information of various scope </td>
        <td rowspan="3">display-group</td>
        <td>name</td>
        <td rowspan="3">update</td>
	</tr>
    <tr>
        <td>icon</td>
	</tr>
    <tr>
        <td>description</td>
	</tr>
    <tr>
        <td>display-${self/member/admin}</td>
        <td>Member.title</td>
        <td>update the member</td>
	</tr>
    <tr>
		<td rowspan="3">Asset information</td>
        <td rowspan="3">asset</td>
        <td>total_equity</td>
        <td>shareholding capacity</td>
	</tr>
    <tr>
        <td>Member.shares</td>
        <td>shares hold by each member</td>
	</tr>
    <tr>
        <td>token_contract</td>
        <td>update the token contract</td>
	</tr>
    <tr>
		<td rowspan="2">Conclusion rules</td>
        <td>conclusion</td>
        <td>conclusion</td>
        <td>add/remove the conclusion category list</td>
	</tr>
    <tr>
        <td>conclusion-${category}</td>
        <td>conclusion_category</td>
        <td>update the rule for each category</td>
	</tr>
    <tr>
		<td rowspan="4">The IT configure</td>
        <td rowspan="4">it</td>
        <td>ood_list</td>
        <td>update the ood list running for group</td>
	</tr>
    <tr>
        <td>history_block_max</td>
        <td>max count of reserved blocks</td>
	</tr>
    <tr>
        <td>history_block_lifespan</td>
        <td>max milliseconds to reserved blocks</td>
	</tr>
    <tr>
        <td>consensus_interval</td>
        <td>max period for consensus established</td>
	</tr>
    <tr>
        <td>Custom permissions</td>
        <td>custom-${any}</td>
        <td></td>
        <td>defined by the DecAPP</td>
	</tr>
</table>

For example:

We require that more than 2 admins agree to modify the member title, or the member can modify it himself.

We can set the rules:

```
display-member=3
display-self=7
```

And assign the permissions:

```
display-member=2 // set to every admins, less than the rule, and greater than rule/2
display-self=9 // set to every member, greater than the rule
```

	And, we can define the configure for each DecAPP on `Conclusion`, `IT`, `Custom`.
	
4. Group generate
	
	* Generate the group object
	* Send to every members and admins
	* Collect all signature from each member and admin, So, no one can join another to a group without notify
	* Push the object to MetaChain
	
	And we can push the object to MetaChain first, and all the admins/members post the signature to MetaChain. Of course, that would cost more commission.
	
5. Update the Group

	* Someone generate a new Group object, and send to others
	* Collect the signatures from members and admins in old version, and new members/admins from new version, ensure the new one has been notified.
	* Check all updated fields collected enough signatures.
		* Yes: update the latest version
		* No: Collect signatures continue
		
	We can collect the signatures off-chain to reduce the commission.
	
6. Cost

	There is commission for every on-chain operation, so we must reduce the frequency, I think there are some suggestion:
	* Save information on the chain as little as possible.
	* Delay submit the request, and wait more updates to submit together.
	* Don't submit updates from anotherone, to avoid someone high-frequency submission irresponsibly.

**Comment by waterflier** [comment]
The Group used in the real scene is really not simple, very deep thinking!

I have two suggestions:
1. Is there an easier way? Is it really necessary to put so much functionality in a standard Group? Should we implement these functions through an extensible design
2. About on-chain cost , I support the "initiator" do on-chain, which is simple and easy to understand.

**Reply by streetycat** [reply]
> The Group used in the real scene is really not simple, very deep thinking!
> 
> I have two suggestions:
> 
> 1. Is there an easier way? Is it really necessary to put so much functionality in a standard Group? Should we implement these functions through an extensible design
> 2. About on-chain cost , I support the "initiator" do on-chain, which is simple and easy to understand.

Good idea, There are only some base infomation and used in `Stack` should be implement in the standard Group, I think only the fields related to `Members` will be reserved, It seems like below:

<table>
	<tr>
		<th>Category resume</th>
		<th>Category</th>
		<th>Field</th>
		<th>Notes</th>
	</tr>
    <tr>
		<td rowspan="2">Member list</td>
        <td>admin</td>
        <td>admins</td>
        <td>add/remove</td>
	</tr>
    <tr>
        <td>member</td>
        <td>members</td>
        <td>add/remove</td>
	</tr>
    <tr>
	<td rowspan="4">Display information of various scope </td>
        <td rowspan="3">display-group</td>
        <td>name</td>
        <td rowspan="3">update</td>
	</tr>
    <tr>
        <td>icon</td>
	</tr>
    <tr>
        <td>description</td>
    </tr>
    <tr>
        <td>display-${self/member/admin}</td>
        <td>Member.title</td>
        <td>update the member</td>
   </tr>
   <tr>
		<td>OOD list</td>
        <td>ood</td>
        <td>ood_list</td>
        <td>add/remove</td>
	</tr>
</table>

And the others are also common functions, They can implement in another way:

<table>
	<tr>
		<td>Asset information</td>
		<td rowspan="3">MetaChain as other Tx</td>
	</tr>
	<tr>
		<td>Conclusion</td>
	</tr>
	<tr>
		<td>Permissions </td>
	</tr>
	<tr>
		<td>it</td>
		<td>config with interface of `cyfs-group`</td>
	</tr>
</table>

And, the details will be designed in future.
